File: apps/web/src/lib/captioning.ts
import type { CaptionGroup, CaptionWord, GroupingSettings } from '../types'

export const defaultGroupingSettings: GroupingSettings = {
  maxWords: 3,
  minDuration: 0.26,
  maxChars: 26,
  pauseThreshold: 0.42,
}

export const timingNudgeStep = 0.05

const minGroupDuration = 0.04

const connectorWords = new Set([
  'а',
  'але',
  'без',
  'бо',
  'в',
  'від',
  'до',
  'з',
  'за',
  'і',
  'й',
  'коли',
  'на',
  'не',
  'ну',
  'та',
  'то',
  'у',
  'це',
  'що',
  'як',
  'якщо',
])

const normalizeWord = (text: string) =>
  text.trim().toLowerCase().replace(/^[.,!?;:"'()[\]{}]+|[.,!?;:"'()[\]{}]+$/g, '')

export const isConnectorWord = (text: string) => connectorWords.has(normalizeWord(text))

const getGroupText = (words: CaptionWord[]) => words.map((word) => word.text).join(' ')

const roundTime = (seconds: number) => Math.round(seconds * 1000) / 1000

export const setGroupTiming = (group: CaptionGroup, start: number, end: number): CaptionGroup => {
  const safeStart = Math.max(0, roundTime(Number.isFinite(start) ? start : group.start))
  const safeEnd = Math.max(safeStart + minGroupDuration, roundTime(Number.isFinite(end) ? end : group.end))

  return {
    ...group,
    start: safeStart,
    end: safeEnd,
  }
}

export const shiftGroupTiming = (group: CaptionGroup, offset: number): CaptionGroup =>
  setGroupTiming(group, group.start + offset, group.end + offset)

export const groupWords = (
  words: CaptionWord[],
  settings: GroupingSettings = defaultGroupingSettings,
): CaptionGroup[] => {
  const groups: CaptionGroup[] = []
  let current: CaptionWord[] = []

  const commit = () => {
    if (!current.length) return
    const first = current[0]
    const last = current[current.length - 1]

    groups.push({
      id: `g_${String(groups.length + 1).padStart(4, '0')}`,
      wordIds: current.map((word) => word.id),
      text: getGroupText(current),
      start: first.start,
      end: last.end,
    })
    current = []
  }

  words.forEach((word) => {
    if (!current.length) {
      current = [word]
      return
    }

    const previous = current[current.length - 1]
    const candidate = [...current, word]
    const candidateText = getGroupText(candidate)
    const duration = word.end - current[0].start
    const gap = word.start - previous.end
    const shouldJoin =
      isConnectorWord(previous.text) || isConnectorWord(word.text) || duration < settings.minDuration
    const exceedsLimits =
      candidate.length > settings.maxWords ||
      candidateText.length > settings.maxChars ||
      gap > settings.pauseThreshold

    if (shouldJoin && !exceedsLimits) {
      current = candidate
      return
    }

    if (current.length === 1 && isConnectorWord(current[0].text) && !exceedsLimits) {
      current = candidate
      return
    }

    commit()
    current = [word]
  })

  commit()
  return groups
}

export const rebuildGroupTiming = (group: CaptionGroup, words: CaptionWord[]): CaptionGroup => {
  const wordMap = new Map(words.map((word) => [word.id, word]))
  const groupWords = group.wordIds.map((id) => wordMap.get(id)).filter(Boolean) as CaptionWord[]

  if (!groupWords.length) return group

  const first = groupWords[0]
  const last = groupWords[groupWords.length - 1]

  return {
    ...group,
    text: getGroupText(groupWords),
    start: first.start,
    end: last.end,
  }
}

export const formatSeconds = (seconds: number) => {
  const minutes = Math.floor(seconds / 60)
  const rest = seconds - minutes * 60
  return `${minutes}:${rest.toFixed(2).padStart(5, '0')}`
}

export const secondsToSrtTime = (seconds: number) => {
  const totalMs = Math.round(seconds * 1000)
  const hours = Math.floor(totalMs / 3_600_000)
  const minutes = Math.floor((totalMs % 3_600_000) / 60_000)
  const secs = Math.floor((totalMs % 60_000) / 1000)
  const millis = totalMs % 1000
  return `${String(hours).padStart(2, '0')}:${String(minutes).padStart(2, '0')}:${String(
    secs,
  ).padStart(2, '0')},${String(millis).padStart(3, '0')}`
}

export const exportSrt = (groups: CaptionGroup[]) =>
  `${groups
    .map((group, index) => {
      const text = group.textOverride?.trim() || group.text
      return `${index + 1}\n${secondsToSrtTime(group.start)} --> ${secondsToSrtTime(group.end)}\n${text}`
    })
    .join('\n\n')}\n`

export const downloadTextFile = (filename: string, content: string) => {
  const blob = new Blob([content], { type: 'text/plain;charset=utf-8' })
  const url = URL.createObjectURL(blob)
  const link = document.createElement('a')
  link.href = url
  link.download = filename
  link.click()
  URL.revokeObjectURL(url)
}

File: apps/api/app/captioning.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable


@dataclass(frozen=True)
class Word:
    id: str
    text: str
    start: float
    end: float
    confidence: float | None = None


@dataclass(frozen=True)
class CaptionGroup:
    id: str
    word_ids: list[str]
    text: str
    start: float
    end: float


@dataclass(frozen=True)
class GroupingSettings:
    max_words: int = 3
    min_duration: float = 0.26
    max_chars: int = 26
    pause_threshold: float = 0.42


def group_words(words: Iterable[Word], settings: GroupingSettings) -> list[CaptionGroup]:
    groups: list[CaptionGroup] = []
    current: list[Word] = []

    def commit() -> None:
        nonlocal current
        if not current:
            return
        first = current[0]
        last = current[-1]
        groups.append(
            CaptionGroup(
                id=f"g_{len(groups) + 1:04d}",
                word_ids=[word.id for word in current],
                text=" ".join(word.text for word in current),
                start=first.start,
                end=last.end,
            )
        )
        current = []

    for word in sorted(words, key=lambda item: (item.start, item.end)):
        if not current:
            current.append(word)
            continue

        candidate = current + [word]
        candidate_text = " ".join(item.text for item in candidate)
        exceeds_hard_limits = len(candidate_text) > settings.max_chars

        if not exceeds_hard_limits:
            current.append(word)
            continue

        commit()
        current = [word]

    commit()
    return groups


def seconds_to_srt_time(seconds: float) -> str:
    milliseconds_total = round(seconds * 1000)
    hours, rest = divmod(milliseconds_total, 3_600_000)
    minutes, rest = divmod(rest, 60_000)
    secs, millis = divmod(rest, 1000)
    return f"{hours:02d}:{minutes:02d}:{secs:02d},{millis:03d}"


def export_srt(groups: Iterable[CaptionGroup]) -> str:
    blocks = []
    for index, group in enumerate(groups, start=1):
        start = seconds_to_srt_time(group.start)
        end = seconds_to_srt_time(group.end)
        blocks.append(f"{index}\n{start} --> {end}\n{group.text}")
    return "\n\n".join(blocks) + "\n"
